docs/design/rubygems.md

```diff
@@ -0,0 +1,284 @@
+The `Gemfile.lock` file is generated by bundler when you run `bundle install` or `bundle update`.
+The only source for gem dependencies to be fetched from is <https://rubygems.org>.
+Git dependencies are specified using a repo URL and pinned to a commit hash.
+
+Bundler always executes the `Gemfile`, which is arbitrary ruby code.
+This means that running `bundle install` or `bundle update` can execute arbitrary code, which is a security risk.
+That's why we **do not use bundler** to download dependencies.
+Instead, as stated above, we parse the `Gemfile.lock` file directly and download the gems from <https://rubygems.org>.
+
+**Note**: parsing `Gemfile.lock` is done via [gemlock-parser](https://github.com/containerbuildsystem/gemlock-parser).
+
+`Gemfile` example:
+
+```ruby
+source 'https://rubygems.org'
+
+gem 'rails', '= 6.1.7'
+
+system('echo "Hello, world!"')
+system('sudo rm -rf /')
+```
+
+### Missing features
+
+Bundler is not pinned as a dependency with version in the `Gemfile.lock` (even if it is pinned in the `Gemfile`).
+But it is important for the same version of bundler to be installable and used for resolving dependencies.
+Using the bundler from the build image usually does not fit.
+
+## Dependencies
+
+Ruby supports three types of dependencies
+
+### Gem dependencies
+
+Regular gem dependencies located at source url, ~~usually~~ in our case always <https://rubygems.org>.
+Each gem can be accessed by its name and version - rubygems.org/gems/`<name>`-`<version>`.gem
+
+Example of a gem dependency in the `Gemfile.lock` file:
+
+```ruby
+GEM
+  remote: https://rubygems.org/
+  specs:
+    ...
+    rails (6.1.4)
+      # transitive dependencies
+      actioncable (= 6.1.4)
+      actionmailbox (= 6.1.4)
+      actionmailer (= 6.1.4)
+      actionpack (= 6.1.4)
+      actiontext (= 6.1.4)
+      actionview (= 6.1.4)
+      activejob (= 6.1.4)
+      activemodel (= 6.1.4)
+      activerecord (= 6.1.4)
+      activestorage (= 6.1.4)
+      activesupport (= 6.1.4)
+      bundler (>= 1.15.0)
+      railties (= 6.1.4)
+      sprockets-rails (>= 2.0.0)
+      ...
+```
+
+### Git dependencies
+
+Example of a git dependency in the `Gemfile.lock` file:
+
+```ruby
+...
+GIT
+  remote: https://github.com/porta.git
+  revision: 779beabd653afcd03c4468e0a69dc043f3bbb748
+  branch: main
+  specs:
+    porta (2.14.1)
+...
+```
+
+Bundler uses this format to cache git repositories:
+
+[https://github.com/rubygems/.../bundler/source/git.rb#L139](https://github.com/rubygems/rubygems/blob/3da9b1dda0824d1d770780352bb1d3f287cb2df5/bundler/lib/bundler/source/git.rb#L139>)
+
+Renaming this directory, or changing the commit hash, will cause bundler to re-download the repository -> cache invalidation -> the build will fail.
+
+### Path dependencies
+
+Recursive copy of the gem directory to the `vendor/cache` directory - but it does not work, if the path is within the project root.
```

Comment:
"Does not work" in what sense? Bundler doesn't copy it to `vendor/cache` because it considers it unnecessary? The PATH dependency can be installed just fine even if it's not in the cache, right?

(this would also make more sense in the Vendoring Solution section)